### ISSUE TITLE
Questions: Can I use StoreKit 2 API?

### ISSUE THREAD

**Issue by Harween64** [xamarin/xamarin-macios#16893]
Hello all,

Everything is in the title.

Is it possible to use StoreKit 2 API with Xamarin iOS or .Net 7?
If yes, how?
If no, is there any plan to enable this in the future?

Thank you

**Comment by rolfbjarne** [comment]
It looks like the StoreKit 2 API is Swift-only, and unfortunately we don't support Swift-only API yet.

**Comment by dalexsoto** [comment]
As of today, the only way to use a swift library is via a proxy library, the general idea is described here https://learn.microsoft.com/en-us/xamarin/ios/platform/binding-swift/ basically you would need to expose the StoreKit 2 Apis you need via a Swift glue layer that exposes the method via the [swift bridge header](https://developer.apple.com/documentation/swift/importing-swift-into-objective-c) and then creating a C# binding project for this.

**Comment by thefex** [comment]
Like @dalexsoto said.

Take a look at my project, you can base on it and extend it based on your needs: 
https://github.com/thefex/Xamarin.iOS.StoreKit2Sample 

There is implementation of one method responsible of showing manage storekit subscriptions. 

You just add appropriate methods in Swift (via Xcode) marked with @objc attribute, then follow Readme - howto build xcframework, build bindings library and consume via nuget. 

Take a note that asynchronous swift methods are crashing on runtime due to the following issue: 
https://github.com/xamarin/xamarin-macios/issues/18562
**on simulator (it works on real device)**

**Comment by taublast** [comment]
@thefex Thanks, this helped me to create a wrapper for some of their new storekit api https://github.com/taublast/AppoMobi.Xamarin.ExternalPurchaseLink, one could use that as another example too, iPhone and simulator are both supported.

**Comment by Harween64** [comment]
@thefex ans @taublast  Thanks ! I will take a look.

**Comment by RamType0** [comment]
As of iOS18.0, we always need to make proxy library by myself to support in app purchase, isn't it?

**Comment by vecalion** [comment]
@rolfbjarne, could you kindly share if there are any updates regarding this issue? With the deprecation of StoreKit 1, it’s becoming increasingly frustrating for us to choose the right development path without knowing Microsoft's plans for this issue.

**Comment by taublast** [comment]
> As of iOS18.0, we always need to make proxy library by myself to support in app purchase, isn't it?

It's not difficult to just bind methods you need from StoreKit 2, the answer is yes.

**Comment by thefex** [comment]
> @rolfbjarne, could you kindly share if there are any updates regarding this issue? With the deprecation of StoreKit 1, it’s becoming increasingly frustrating for us to choose the right development path without knowing Microsoft's plans for this issue.

Check my sample above (link to github repository), I bet you need just few methods and you could do that alone in ~20 minutes.

**Comment by RamType0** [comment]
@rolfbjarne We need response from Microsoft.

**Comment by RamType0** [comment]
> [@RamType0](https://github.com/RamType0) Swift interop is tracked here [dotnet/runtime#95638](https://github.com/dotnet/runtime/issues/95638) we are not ready at this time to provide more updates at the moment other that it is being worked on, unfortunately this is not trivial to do and requires runtime support for swift before we can write any c# bindings to anything swift

Some people is posting their own StoreKit2 bindings.
Why Microsoft cannot provide these kind of bindings?

**Comment by thefex** [comment]
> > [@RamType0](https://github.com/RamType0) Swift interop is tracked here [dotnet/runtime#95638](https://github.com/dotnet/runtime/issues/95638) we are not ready at this time to provide more updates at the moment other that it is being worked on, unfortunately this is not trivial to do and requires runtime support for swift before we can write any c# bindings to anything swift
> 
> Some people is posting their own StoreKit2 bindings.
> Why Microsoft cannot provide these kind of bindings?
*non MS employee answer 

Because that requires manual steps -> "writing" obj-c code that bridges Swift into Objective C world.
Then running .NET-ios bindings generator over the generated Obj-C compatible framework.

I guess MS needs to either natively support swift or write their own Obj-C Swift-compatibility generator to easily deploy and maintain those bindings

**Comment by NatElkins** [comment]
I believe this is the specific branch where the Swift interop work is happening. https://github.com/dotnet/runtimelab/tree/feature/swift-bindings

**Comment by anose001** [comment]
Revisiting this issue. Microsoft [indicates](https://learn.microsoft.com/en-us/samples/dotnet/maui-samples/cross-platform-billing-service/) using StoreKit 1 for .net ios. Does anybody know how much longer StoreKit 1 will be available to use and whether there is any risk of Apple removing StoreKit 1 and forcing developers to use StoreKit 2 which currently does not have support for dotnet ios?

**Comment by thefex** [comment]
> Revisiting this issue. Microsoft [indicates](https://learn.microsoft.com/en-us/samples/dotnet/maui-samples/cross-platform-billing-service/) using StoreKit 1 for .net ios. Does anybody know how much longer StoreKit 1 will be available to use and whether there is any risk of Apple removing StoreKit 1 and forcing developers to use StoreKit 2 which currently does not have support for dotnet ios?

You should use StoreKit2 and migrate into it, new features are storekit2. But I personally just moved into revenuecat - wraps storekit, makes payment integration and their pricing model (billed based on your revenue) is suitable.

**Comment by anose001** [comment]
@thefex , are you on dotnet ios? If so I am curious to learn more how you went down that route.

**Comment by thefex** [comment]
> @thefex , are you on dotnet ios? If so I am curious to learn more how you went down that route.

Yup. Just use .net-ios bindings for a service that offers payment integration (i.e. revenuecat is one of many), i already created some bindings - will update it somewhere in future - https://www.nuget.org/packages/RevenueCat.DotNet.Bindings

But its not hard to do bindings on your own.

Besides... payments are not easy to handle all edge cases and never ending new features/api changes

**Comment by anose001** [comment]
Thank you for sharing that, it looks like the path of least resistance with how everything is standing right now. What has your experience been with breaking changes and version upgrades? 
Would you consider open-sourcing the repository? I would like to take a peak at it and reference it as a starting point

**Comment by thefex** [comment]
> Thank you for sharing that, it looks like the path of least resistance with how everything is standing right now. What has your experience been with breaking changes and version upgrades? 
> Would you consider open-sourcing the repository? I would like to take a peak at it and reference it as a starting point

Sure, join the .net discord and DM me, i got same nick
